flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/autoscaler/JobVertexScaler.java

```diff
@@ -0,0 +1,264 @@
+            Configuration conf,
+            Map<ScalingMetric, EvaluatedScalingMetric> evaluatedMetrics,
+            SortedMap<Instant, ScalingSummary> history,
+            int currentParallelism,
+            int newParallelism) {
+
+        // If we don't have past scaling actions for this vertex, there is nothing to do
+        if (history.isEmpty()) {
+            return false;
+        }
+
+        boolean scaleUp = currentParallelism < newParallelism;
+
+        return detectImmediateScaleDownAfterScaleUp(vertex, conf, history, scaleUp)
+                || detectIneffectiveScaleup(vertex, conf, evaluatedMetrics, history, scaleUp);
+    }
+
+    private boolean detectImmediateScaleDownAfterScaleUp(
+            JobVertexID vertex,
+            Configuration conf,
+            SortedMap<Instant, ScalingSummary> history,
+            boolean scaleUp) {
+
+        // Does not apply to scale up operations
+        if (scaleUp) {
+            return false;
+        }
+
+        var lastScalingTs = history.lastKey();
+        var lastSummary = history.get(lastScalingTs);
+
+        boolean lastScaleUp = lastSummary.getNewParallelism() > lastSummary.getCurrentParallelism();
+        if (!lastScaleUp) {
+            return false;
+        }
+
+        var gracePeriod = conf.get(SCALE_UP_GRACE_PERIOD);
+        if (lastScalingTs.plus(gracePeriod).isAfter(clock.instant())) {
+            LOG.info(
+                    "Skipping immediate scale down after scale up within grace period for {}",
+                    vertex);
+            return true;
+        } else {
+            return false;
+        }
+    }
+
+    private boolean detectIneffectiveScaleup(
+            JobVertexID vertex,
+            Configuration conf,
+            Map<ScalingMetric, EvaluatedScalingMetric> evaluatedMetrics,
+            SortedMap<Instant, ScalingSummary> history,
+            boolean scaleUp) {
+
+        // Does not apply to scale down operations
+        if (!scaleUp) {
+            return false;
+        }
+
+        var lastScalingTs = history.lastKey();
+        var lastSummary = history.get(lastScalingTs);
+
+        boolean lastScaleUp = lastSummary.getNewParallelism() > lastSummary.getCurrentParallelism();
+        if (!lastScaleUp) {
+            return false;
+        }
+
+        double lastProcRate = lastSummary.getMetrics().get(TRUE_PROCESSING_RATE).getAverage();
+        double lastExpectedProcRate =
+                lastSummary.getMetrics().get(EXPECTED_PROCESSING_RATE).getCurrent();
+        var currentProcRate = evaluatedMetrics.get(TRUE_PROCESSING_RATE).getAverage();
+
+        double expectedIncrease = lastExpectedProcRate / lastProcRate - 1;
+        double actualIncrease = currentProcRate / lastProcRate - 1;
+
+        boolean withinEffectiveThreshold =
+                (expectedIncrease - actualIncrease) / expectedIncrease
+                        < (1 - conf.get(AutoScalerOptions.SCALING_EFFECTIVENESS_THRESHOLD));
```

Comment:
Maybe adding comments would help to explain what your formula does beyond this.

Reply:
The intention here is not to compare against the percentage change (what your code does) because this will make this unusuable for large jobs when the expected change may actually be small. Going from parallelism 1000 -> 1010 may be a 1% expected change and that is completely efficient. 

What we actually want is if we expected a 1.01 (1%) increase and we only got (1.0001) we need to mark this uneffective. So we actually compare the change itself so we have to subtract 1.

This is not immediately obvious and I spent some time thinking about it. I think this is the only correct way that is parallelism agnostic.

Reply:
Oh right, we want to compare the percentage on the delta. That makes a lot of sense. So my code above should have been:

```
delta_old = expected_rate - old_rate
delta_new = actual_rate - expected_rate
percentage_delta_change = delta_new / delta_old - 1;
if (percentage_delta_change < conf.get(AutoScalerOptions.SCALING_EFFECTIVENESS_THRESHOLD)) {
  return false;
} else {
   // uneffective scaling detected
}
```

Reply:
I think your approach is still different, as you compare the percentages on the total rate whereas I compare the deltas. Result is probably similar but I think the above is what we actually want to do to measure the effectiveness most closely.